tests/common/helpers/parallel_utils.py

```diff
@@ -1,5 +1,180 @@
+import fcntl
+import logging
+import multiprocessing
+from datetime import datetime
+from enum import Enum
+from typing import Tuple
+
 from tests.common import config_reload
+from tests.common.helpers.assertions import pytest_assert as pt_assert
+from tests.common.utilities import wait_until
+
+logger = logging.getLogger(__name__)
 
 
 def config_reload_parallel_compatible(node, results, *args, **kwargs):
     return config_reload(node, *args, **kwargs)
+
+
+def is_initial_checks_active(request):
+    parallel_state_file = request.config.getoption("--parallel_state_file")
+    try:
+        current_parallel_status = read_last_line_from_file(parallel_state_file).split(',')[1]
+        logger.debug("current parallel status: {}".format(current_parallel_status))
+        return current_parallel_status.endswith("started")
+    except FileNotFoundError:
+        logger.info("File {} not found".format(parallel_state_file))
+        return False
+
+
+def read_last_line_from_file(file):
+    with open(file, 'r') as f:
+        fcntl.flock(f, fcntl.LOCK_EX)
+        try:
+            return f.readlines()[-1].strip()
+        except FileNotFoundError:
+            logger.error("File {} not found".format(file))
+            raise FileNotFoundError
+        finally:
+            fcntl.flock(f, fcntl.LOCK_UN)
+
+
+class InitialCheckStatus(Enum):
+    IDLE = "idle"
+    SETUP_STARTED = "setup_started"
+    SETUP_COMPLETED = "setup_completed"
+    TESTS_COMPLETED = "tests_completed"
+    TEARDOWN_STARTED = "teardown_started"
+    TEARDOWN_COMPLETED = "teardown_completed"
+    SANITY_CHECK_FAILED = "sanity_check_failed"
```

Comment:
We use a state machine to control the progress of the test for the leader and the followers. Here's a diagram for better understanding the state changes:
![parallel run state](https://github.com/user-attachments/assets/4c9c45a6-9bb8-4eb1-94b9-892617334a89)